File: ui/src/lib/homography.ts
// Computes a 2D affine "rectification" from the scene's labeled reference
// strokes. Returns null when the labels don't provide enough constraint
// (the UI shows "noch nicht genug Bezüge" in that case).
//
// First-cut: affine-only. A horizontal reference stroke + a vertical
// reference stroke pin scale, rotation, and a uniform shear. Full
// perspective rectification (8-DOF homography) needs 4 corner-region
// strokes — accepted as a future upgrade in spec/annotation-tool.md §7.2.
//
// All inputs/outputs are in raw pixel coordinates; mm values are the
// real-world lengths the user labeled.

import type { DimensionedDistanceLabel, Label, Point } from '../api/types';

export interface Affine {
  // x_world = a*x_px + c*y_px + tx
  // y_world = b*x_px + d*y_px + ty
  a: number; b: number; c: number; d: number; tx: number; ty: number;
}

export interface RectificationResult {
  affine: Affine;
  /** 3x3 row-major for storage; bottom row [0,0,1] for an affine. */
  matrix: number[][];
  /** Stroke ids used to compute the transform. */
  computed_from: string[];
  /** Bounding-box size of the rectified image, in scaled output pixels. */
  rectified_size_px: [number, number];
  /** Top-left corner of the warped image in scaled-output pixel coords. */
  rectified_offset_px: [number, number];
  /** Pixels-per-mm scale applied to fit `target_max_dim` (display only). */
  display_scale: number;
  /** Sum of per-stroke residuals after fitting, in scaled-output pixels. */
  rms_residual_px: number;
  /** 'ok' | 'insufficient_references' | 'degenerate' */
  status: 'ok' | 'insufficient_references' | 'degenerate';
  reason?: string;
}

export interface RectifyOptions {
  /** Max dimension of the rectified output (so output fits a UI pane). */
  target_max_dim?: number;
  imageSize: [number, number];
}

const DEFAULT_TARGET_MAX = 1200;

function isDimDistance(l: Label): l is DimensionedDistanceLabel {
  return l.type === 'dimensioned_distance';
}

function strokeMmDir(s: DimensionedDistanceLabel): { px: Point; mm: number } | null {
  const v: Point = [s.geometry.end[0] - s.geometry.start[0], s.geometry.end[1] - s.geometry.start[1]];
  const mm = s.attributes.value_mm;
  if (mm == null) return null;
  return { px: v, mm };
}

function pickLongest(
  refs: DimensionedDistanceLabel[],
  isAxis: (s: DimensionedDistanceLabel) => boolean,
): DimensionedDistanceLabel | null {
  let best: DimensionedDistanceLabel | null = null;
  let bestLen = -1;
  for (const s of refs) {
    if (!isAxis(s)) continue;
    const v = strokeMmDir(s);
    if (!v) continue;
    const len = Math.hypot(v.px[0], v.px[1]);
    if (len > bestLen) {
      best = s;
      bestLen = len;
    }
  }
  return best;
}

function isHorizontal(s: DimensionedDistanceLabel): boolean {
  const o = s.attributes.target_orientation;
  return o === 'horizontal' || (typeof o === 'string' && /^angle_deg:0(\.0+)?$/.test(o));
}

function isVertical(s: DimensionedDistanceLabel): boolean {
  const o = s.attributes.target_orientation;
  return o === 'vertical' || (typeof o === 'string' && /^angle_deg:90(\.0+)?$/.test(o));
}

/**
 * Compute the affine that maps image pixels → rectified output pixels using
 * the labeled reference strokes. Currently: needs ≥1 horizontal + ≥1
 * vertical reference (both `is_reference: true` + non-null `value_mm`).
 * Picks the longest of each axis; returns 'insufficient_references' otherwise.
 */
export function computeRectification(
  labels: Label[],
  opts: RectifyOptions,
): RectificationResult {
  const refs = labels
    .filter(isDimDistance)
    .filter((s) => s.attributes.is_reference && s.attributes.value_mm != null);

  const H = pickLongest(refs, isHorizontal);
  const V = pickLongest(refs, isVertical);

  if (!H || !V) {
    return {
      affine: { a: 1, b: 0, c: 0, d: 1, tx: 0, ty: 0 },
      matrix: [[1, 0, 0], [0, 1, 0], [0, 0, 1]],
      computed_from: [],
      rectified_size_px: opts.imageSize,
      rectified_offset_px: [0, 0],
      display_scale: 1,
      rms_residual_px: 0,
      status: 'insufficient_references',
      reason: !H && !V
        ? 'Mindestens 1 horizontale + 1 vertikale Referenz-Strecke benötigt.'
        : !H
          ? 'Es fehlt eine horizontale Referenz-Strecke (is_reference + value_mm + target_orientation=horizontal).'
          : 'Es fehlt eine vertikale Referenz-Strecke.',
    };
  }

  const vh = strokeMmDir(H)!;
  const vv = strokeMmDir(V)!;
  const [px, py] = vh.px;
  const [qx, qy] = vv.px;
  const Lh = vh.mm;
  const Lv = vv.mm;

  // Solve M * [px qx; py qy] = [Lh 0; 0 Lv].
  const det = px * qy - qx * py;
  if (Math.abs(det) < 1e-6) {
    return {
      affine: { a: 1, b: 0, c: 0, d: 1, tx: 0, ty: 0 },
      matrix: [[1, 0, 0], [0, 1, 0], [0, 0, 1]],
      computed_from: [H.id, V.id],
      rectified_size_px: opts.imageSize,
      rectified_offset_px: [0, 0],
      display_scale: 1,
      rms_residual_px: 0,
      status: 'degenerate',
      reason: 'Horizontale und vertikale Referenz sind nicht linear unabhängig — Strecken stehen fast parallel zueinander.',
    };
  }

  // World-coords matrix M (px → mm). y axis: SVG y grows downward, world y
  // grows downward too (consistent with the on-page drawing orientation),
  // so we keep the raw sign.
  const m11 = (Lh * qy) / det;
  const m12 = (-Lh * qx) / det;
  const m21 = (-Lv * py) / det;
  const m22 = (Lv * px) / det;

  // Compute the bbox in mm of all four image corners after applying (M, 0).
  // Then translate so the top-left mm corner sits at (0, 0), and scale to
  // fit the requested output max dim.
  const [imgW, imgH] = opts.imageSize;
  const corners: Point[] = [[0, 0], [imgW, 0], [imgW, imgH], [0, imgH]];
  const mmCorners = corners.map(([x, y]) => [m11 * x + m12 * y, m21 * x + m22 * y] as Point);
  let minX = Infinity, minY = Infinity, maxX = -Infinity, maxY = -Infinity;
  for (const [x, y] of mmCorners) {
    if (x < minX) minX = x;
    if (y < minY) minY = y;
    if (x > maxX) maxX = x;
    if (y > maxY) maxY = y;
  }
  const mmW = maxX - minX;
  const mmH = maxY - minY;
  const target = opts.target_max_dim ?? DEFAULT_TARGET_MAX;
  const scale = target / Math.max(mmW, mmH);

  // Final affine: M' = scale * (M, with translation that maps (minX, minY)
  // to (0, 0)).
  const a = m11 * scale;
  const c = m12 * scale;
  const b = m21 * scale;
  const d = m22 * scale;
  const tx = -minX * scale;
  const ty = -minY * scale;

  // Residual estimate: project both refs through the affine, compare against
  // expected world lengths.
  const projLen = (p: Point) => {
    const w: Point = [a * p[0] + c * p[1] + tx, b * p[0] + d * p[1] + ty];
    return w;
  };
  const wH0 = projLen(H.geometry.start);
  const wH1 = projLen(H.geometry.end);
  const wV0 = projLen(V.geometry.start);
  const wV1 = projLen(V.geometry.end);
  const horizScaledLen = Math.hypot(wH1[0] - wH0[0], wH1[1] - wH0[1]);
  const vertScaledLen = Math.hypot(wV1[0] - wV0[0], wV1[1] - wV0[1]);
  const expHoriz = Lh * scale;
  const expVert = Lv * scale;
  const rms = Math.sqrt(
    ((horizScaledLen - expHoriz) ** 2 + (vertScaledLen - expVert) ** 2) / 2,
  );

  return {
    affine: { a, b, c, d, tx, ty },
    matrix: [[a, c, tx], [b, d, ty], [0, 0, 1]],
    computed_from: [H.id, V.id],
    rectified_size_px: [Math.round(mmW * scale), Math.round(mmH * scale)],
    rectified_offset_px: [0, 0],
    display_scale: scale,
    rms_residual_px: rms,
    status: 'ok',
  };
}

/** Apply an affine to a point. */
export function applyAffine(A: Affine, p: Point): Point {
  return [A.a * p[0] + A.c * p[1] + A.tx, A.b * p[0] + A.d * p[1] + A.ty];
}

/** Apply an affine to a whole label by walking the relevant geometry fields. */
export function rectifyLabel(A: Affine, l: Label): Label {
  switch (l.type) {
    case 'wall':
    case 'dimensioned_distance':
      return {
        ...l,
        geometry: { start: applyAffine(A, l.geometry.start), end: applyAffine(A, l.geometry.end) },
      } as Label;
    case 'dimension_number':
      return {
        ...l,
        geometry: {
          anchor: l.geometry.anchor ? applyAffine(A, l.geometry.anchor) : undefined,
          bbox: l.geometry.bbox
            ? (l.geometry.bbox.map((p) => applyAffine(A, p)) as [Point, Point, Point, Point])
            : undefined,
        },
      } as Label;
    case 'floorplan_opening':
      return {
        ...l,
        geometry: {
          quad: l.geometry.quad.map((p) => applyAffine(A, p)) as [Point, Point, Point, Point],
        },
      } as Label;
    case 'view_opening':
      return {
        ...l,
        geometry: {
          top_edge: l.geometry.top_edge.map((p) => applyAffine(A, p)) as Point[],
          bottom_edge: l.geometry.bottom_edge.map((p) => applyAffine(A, p)) as Point[],
        },
      } as Label;
    case 'component_line':
      return {
        ...l,
        geometry: { polyline: l.geometry.polyline.map((p) => applyAffine(A, p)) as Point[] },
      } as Label;
    case 'height_mark':
      return { ...l, geometry: { anchor: applyAffine(A, l.geometry.anchor) } } as Label;
  }
}

File: api/homography.py
"""R4 — server-side homography + rectification.

Mirrors ui/src/lib/homography.ts (affine-only, longest-H × longest-V).
Used by /exports/{key}/{file}/preview to compute the rectified image
+ a transformed copy of every label for Set B.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any


@dataclass
class Affine:
    """x_w = a*x_px + c*y_px + tx; y_w = b*x_px + d*y_px + ty"""
    a: float
    b: float
    c: float
    d: float
    tx: float
    ty: float


@dataclass
class Rectification:
    affine: Affine
    matrix: list[list[float]]
    computed_from: list[str]
    rectified_size_px: tuple[int, int]
    display_scale: float
    rms_residual_px: float
    status: str  # 'ok' | 'insufficient_references' | 'degenerate'
    reason: str | None = None
    # Issue #26: True when this calibration used a single reference dim
    # and derived the second-axis scale from the same px-per-mm under the
    # square-pixel (isotropic) assumption — valid for axis-aligned
    # orthographic drawings (German Ansicht/Schnitt). The harness
    # vision-LLM makes the axis-aligned judgement and opts in via
    # `assume_isotropic`; the engine just honours it and records it here.
    single_ref_assumed_isotropic: bool = False


def _orient_is_horiz(o: Any) -> bool:
    if o == "horizontal":
        return True
    if isinstance(o, str) and o.startswith("angle_deg:"):
        try:
            v = float(o.split(":", 1)[1])
            return abs(v) < 0.5
        except ValueError:
            return False
    return False


def _orient_is_vert(o: Any) -> bool:
    if o == "vertical":
        return True
    if isinstance(o, str) and o.startswith("angle_deg:"):
        try:
            v = float(o.split(":", 1)[1])
            return abs(v - 90.0) < 0.5
        except ValueError:
            return False
    return False


def _pick_longest(refs: list[dict], axis: str) -> dict | None:
    best, best_len = None, -1.0
    for s in refs:
        attrs = s.get("attributes") or {}
        geom = s.get("geometry") or {}
        if axis == "h" and not _orient_is_horiz(attrs.get("target_orientation")):
            continue
        if axis == "v" and not _orient_is_vert(attrs.get("target_orientation")):
            continue
        if attrs.get("value_mm") is None:
            continue
        try:
            sx, sy = geom["start"]
            ex, ey = geom["end"]
        except (KeyError, TypeError, ValueError):
            continue
        vx, vy = ex - sx, ey - sy
        length = (vx * vx + vy * vy) ** 0.5
        if length > best_len:
            best, best_len = s, length
    return best


def compute_rectification(
    labels: list[dict],
    image_size: tuple[int, int],
    target_max_dim: int = 1200,
    assume_isotropic: bool = False,
) -> Rectification:
    """Affine fit from longest H + longest V is_reference dim_distances.
    Returns a typed result with status='ok'/'insufficient_references'/'degenerate'."""
    refs = [
        l for l in labels
        if l.get("type") == "dimensioned_distance"
        and (l.get("attributes") or {}).get("is_reference") is True
        and (l.get("attributes") or {}).get("value_mm") is not None
    ]
    H = _pick_longest(refs, "h")
    V = _pick_longest(refs, "v")
    identity = Affine(1, 0, 0, 1, 0, 0)

    # Issue #26 — single-ref isotropic calibration.
    # Axis-aligned orthographic drawings (all German Ansicht/Schnitt)
    # have square pixels, so one reliable reference dim fixes the scale
    # on BOTH axes: px-per-mm is identical horizontally and vertically.
    # When the caller (the harness vision-LLM, having judged the drawing
    # orthographic) opts in via assume_isotropic, synthesise the missing
    # axis from the present one — a unit-aspect partner rotated 90 deg
    # with the same mm-per-px — so the existing two-ref affine math runs
    # unchanged and yields an isotropic transform instead of erroring.
    single_ref_isotropic = False
    if assume_isotropic and (H is None) != (V is None):
        present = H if H is not None else V
        pg = present["geometry"]
        psx, psy = pg["start"]
        pex, pey = pg["end"]
        pvx, pvy = pex - psx, pey - psy
        plen_px = (pvx * pvx + pvy * pvy) ** 0.5
        plen_mm = float(present["attributes"]["value_mm"])
        if plen_px >= 1e-6 and plen_mm > 0:
            # Rotate the present pixel vector +90 deg; same length in px
            # AND in mm (square pixels => identical px-per-mm scale).
            rot_vx, rot_vy = -pvy, pvx
            synth = {
                "id": "synthetic_isotropic",
                "geometry": {
                    "start": [psx, psy],
                    "end": [psx + rot_vx, psy + rot_vy],
                },
                "attributes": {"value_mm": plen_mm},
            }
            if H is None:
                H = synth
            else:
                V = synth
            single_ref_isotropic = True

    if H is None or V is None:
        reason = (
            "Mindestens 1 horizontale + 1 vertikale Referenz-Strecke benötigt."
            if H is None and V is None
            else ("Es fehlt eine horizontale Referenz-Strecke." if H is None
                  else "Es fehlt eine vertikale Referenz-Strecke.")
        )
        return Rectification(
            affine=identity,
            matrix=[[1, 0, 0], [0, 1, 0], [0, 0, 1]],
            computed_from=[],
            rectified_size_px=image_size,
            display_scale=1.0,
            rms_residual_px=0.0,
            status="insufficient_references",
            reason=reason,
        )
    hx, hy = H["geometry"]["end"][0] - H["geometry"]["start"][0], H["geometry"]["end"][1] - H["geometry"]["start"][1]
    vx, vy = V["geometry"]["end"][0] - V["geometry"]["start"][0], V["geometry"]["end"][1] - V["geometry"]["start"][1]
    Lh = float(H["attributes"]["value_mm"])
    Lv = float(V["attributes"]["value_mm"])

    det = hx * vy - vx * hy
    if abs(det) < 1e-6:
        return Rectification(
            affine=identity,
            matrix=[[1, 0, 0], [0, 1, 0], [0, 0, 1]],
            computed_from=[H["id"], V["id"]],
            rectified_size_px=image_size,
            display_scale=1.0,
            rms_residual_px=0.0,
            status="degenerate",
            reason="Horizontale und vertikale Referenz stehen fast parallel zueinander.",
        )

    m11 = (Lh * vy) / det
    m12 = (-Lh * vx) / det
    m21 = (-Lv * hy) / det
    m22 = (Lv * hx) / det

    img_w, img_h = image_size
    corners = [(0.0, 0.0), (img_w, 0.0), (img_w, img_h), (0.0, img_h)]
    mm_corners = [(m11 * x + m12 * y, m21 * x + m22 * y) for x, y in corners]
    xs = [p[0] for p in mm_corners]
    ys = [p[1] for p in mm_corners]
    min_x, min_y = min(xs), min(ys)
    max_x, max_y = max(xs), max(ys)
    mm_w = max_x - min_x
    mm_h = max_y - min_y
    scale = target_max_dim / max(mm_w, mm_h)

    a = m11 * scale; c = m12 * scale
    b = m21 * scale; d = m22 * scale
    tx = -min_x * scale
    ty = -min_y * scale

    def project(pt: tuple[float, float]) -> tuple[float, float]:
        return a * pt[0] + c * pt[1] + tx, b * pt[0] + d * pt[1] + ty

    wH0 = project((H["geometry"]["start"][0], H["geometry"]["start"][1]))
    wH1 = project((H["geometry"]["end"][0], H["geometry"]["end"][1]))
    wV0 = project((V["geometry"]["start"][0], V["geometry"]["start"][1]))
    wV1 = project((V["geometry"]["end"][0], V["geometry"]["end"][1]))
    horiz_len = ((wH1[0] - wH0[0]) ** 2 + (wH1[1] - wH0[1]) ** 2) ** 0.5
    vert_len  = ((wV1[0] - wV0[0]) ** 2 + (wV1[1] - wV0[1]) ** 2) ** 0.5
    exp_h = Lh * scale
    exp_v = Lv * scale
    rms = (((horiz_len - exp_h) ** 2 + (vert_len - exp_v) ** 2) / 2) ** 0.5

    return Rectification(
        affine=Affine(a, b, c, d, tx, ty),
        matrix=[[a, c, tx], [b, d, ty], [0, 0, 1]],
        computed_from=[H["id"], V["id"]],
        rectified_size_px=(round(mm_w * scale), round(mm_h * scale)),
        display_scale=scale,
        rms_residual_px=rms,
        status="ok",
        single_ref_assumed_isotropic=single_ref_isotropic,
    )


def apply_affine(A: Affine, x: float, y: float) -> tuple[float, float]:
    return A.a * x + A.c * y + A.tx, A.b * x + A.d * y + A.ty


def transform_label(A: Affine, l: dict) -> dict:
    """Return a deep-ish copy of `l` with geometry coordinates pushed
    through the affine. Status / attributes are preserved verbatim.
    Caller passes Set-B labels only."""
    out = {**l, "geometry": {**l.get("geometry", {})}}
    g = out["geometry"]
    t = l.get("type")
    if t in ("wall", "dimensioned_distance") and "start" in g and "end" in g:
        g["start"] = list(apply_affine(A, *g["start"]))
        g["end"]   = list(apply_affine(A, *g["end"]))
    elif t == "floorplan_opening" and "quad" in g:
        g["quad"] = [list(apply_affine(A, x, y)) for x, y in g["quad"]]
    elif t == "view_opening":
        if "shape" in g and g["shape"] == "circle" and "center" in g:
            g["center"] = list(apply_affine(A, *g["center"]))
        elif "shape" in g and g["shape"] == "polygon" and "polygon" in g:
            g["polygon"] = [list(apply_affine(A, x, y)) for x, y in g["polygon"]]
        else:
            if "top_edge" in g:
                g["top_edge"] = [list(apply_affine(A, x, y)) for x, y in g["top_edge"]]
            if "bottom_edge" in g:
                g["bottom_edge"] = [list(apply_affine(A, x, y)) for x, y in g["bottom_edge"]]
    elif t == "component_line" and "polyline" in g:
        g["polyline"] = [list(apply_affine(A, x, y)) for x, y in g["polyline"]]
    elif t == "height_mark" and "anchor" in g:
        g["anchor"] = list(apply_affine(A, *g["anchor"]))
    elif t == "dimension_number":
        if "anchor" in g and g["anchor"]:
            g["anchor"] = list(apply_affine(A, *g["anchor"]))
        if "bbox" in g and g["bbox"]:
            g["bbox"] = [list(apply_affine(A, x, y)) for x, y in g["bbox"]]
    return out


def rectify_image(src_path, dst_path, A: Affine, out_size: tuple[int, int]) -> None:
    """Apply the affine via PIL.Image.transform(AFFINE). PIL takes the
    INVERSE affine — given output pixel (u, v) it computes the source
    pixel (a*u + b*v + c, d*u + e*v + f)."""
    from PIL import Image
    det = A.a * A.d - A.c * A.b
    if abs(det) < 1e-9:
        raise ValueError("affine is degenerate; cannot invert")
    ia =  A.d / det
    ib = -A.c / det
    ic = -(ia * A.tx + ib * A.ty)
    id_ = -A.b / det
    ie =  A.a / det
    if_ = -(id_ * A.tx + ie * A.ty)
    with Image.open(src_path) as im:
        im = im.convert("RGB")
        out = im.transform(out_size, Image.AFFINE, (ia, ib, ic, id_, ie, if_), Image.BICUBIC)
        out.save(dst_path, format="JPEG", quality=88)
